proposals/NNNN-conditional-conformances.md

```diff
@@ -0,0 +1,362 @@
+3. All of the disambiguation machinery required at compile time (e.g., to determine whether one conditional conformance is more specialized than another to order them) also needs to implements in the run-time, as part of the dynamic casting machinery. One must also address the possibility of ambiguities occurring at run-time. This is both a sharp increase in the complexity of the system and a potential run-time performance hazard.
+
+For these reasons, this proposal *bans overlapping conformances* entirely. While the resulting system is less flexible than one that allowed overlapping conformances, the gain in simplicity in this potentially-confusing area is well worth the cost. Moreover, this ban follows with existing Swift rules regarding multiple conformances, which prohibit the same type from conforming to the same protocol in two different ways:
+
+```swift
+protocol P { }
+
+struct S : P { }
+extension S : P { } // error: S already conforms to P
+```
+
+### Implied conditional conformances 
+
+Stating conformance to a protocol implicitly states conformances to any of the protocols that it inherits. This is the case in Swift today, although most developers likely don't realize the rules it follows. For example:
+
+```swift
+protocol P { }
+protocol Q : P { }
+protocol R : P { }
+
+struct X1 { }
+struct X2 { }
+struct X3 { }
+
+extension X1: Q { }  // implies conformance to P
+
+extension X2: Q { }  // would imply conformance to P, but...
+extension X2: P { }  // explicitly-stated conformance to P "wins"
+
+extension X3: Q { }  // implies conformance to P
+extension X3: R { }  // also implies conformance to P
+                     // one will "win"; which is unspecified
+```
+
+With conditional conformances, the question of which extension "wins" the implied conformance begins to matter, because the extensions might have different constraints on them. For example:
+
+```swift
+struct X4<T> { }
+
+extension X4: Q where T: Q { }  // implies conformance to P
+extension X4: R where T: R { }  // error: implies overlapping conformance to P
+```
+
+Both of these constrained extensions imply a conformance to `P`, but the actual `P` implied conformances to `P` are overlapping and, therefore, result in an error.
+
+However, in cases where there is a reasonable ordering between the two constrained extensions (i.e., one is more specialized than the other), the less specialized constrained extension should "win" the implied conformance. Continuing the example from above:
+
+```swift
+protocol S: R { }
+
+struct X5<T> { }
+
+extension X5: R where T: R { }  // "wins" implied conformance to P, because
```

Comment:
The first extension "wins" the conformance, but the formatting is confusing. @jckarter has better wording/formatting he'll use to fix this up. Thanks!